iis-wf/iis-wf-affmatching/src/main/java/eu/dnetlib/iis/wf/affmatching/write/BestMatchedAffiliationWithinDocumentPicker.java

```diff
@@ -0,0 +1,78 @@
+package eu.dnetlib.iis.wf.affmatching.write;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+import eu.dnetlib.iis.wf.affmatching.model.MatchedAffiliation;
+import scala.Tuple2;
+
+/**
+ * Picker of best matched affiliation within single document
+ * 
+ * @author madryk
+ */
+public class BestMatchedAffiliationWithinDocumentPicker implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+
+    //------------------------ LOGIC --------------------------
+    
+    /**
+     * Picks best matched affiliation from matched affiliations within single document.<br/>
+     * 
+     * Matched affiliation is considered the best if its {@link MatchedAffiliation#getOrganizationId()}
+     * appeared in all matches more commonly then other organization ids and its
+     * {@link MatchedAffiliation#getMatchStrength()} was highest.
+     */
+    public MatchedAffiliation pickBest(Iterable<MatchedAffiliation> matchedAffs) {
+        
+        Collection<List<MatchedAffiliation>> matchedGroupedByOrgId = groupByOrganizationId(matchedAffs);
+        
+        return matchedGroupedByOrgId.stream()
+            .map(matchesWithSameOrg -> new Tuple2<>(matchesWithSameOrg.size(), fetchMatchWithMaxStrength(matchesWithSameOrg)))
+            .reduce((match1, match2) -> pickBetterMatch(match1, match2))
+            .get()._2;
+        
+    }
+    
+    
+    //------------------------ PRIVATE --------------------------
+    
+    private Collection<List<MatchedAffiliation>> groupByOrganizationId(Iterable<MatchedAffiliation> matchedAffs) {
+        Map<CharSequence, List<MatchedAffiliation>> matchedGroupedByOrgId = Maps.newHashMap();
+        
+        for (MatchedAffiliation matched : matchedAffs) {
+            
+            if (!matchedGroupedByOrgId.containsKey(matched.getOrganizationId())) {
+                matchedGroupedByOrgId.put(matched.getOrganizationId(), Lists.newArrayList(matched));
+            } else {
+                matchedGroupedByOrgId.get(matched.getOrganizationId()).add(matched);
+            }
+            
+        }
+        
+        return matchedGroupedByOrgId.values();
+    }
+    
+    private Tuple2<Integer, MatchedAffiliation> pickBetterMatch(Tuple2<Integer, MatchedAffiliation> first, Tuple2<Integer, MatchedAffiliation> second) {
+        
+        if (first._1 > second._1) {
+            return first;
```

Comment:
It seems this solution is tailored for the current development phase in which we cannot put much trust on `matchStrength`. When we adjust this value to represent match strength accurately we should reimplement it in a way we trust more `matchStrength` not the number of matches. Can you create issue for this so we won't forget about it? It might be quite crucial in the end.

I've got this gut filling there is some paper referring to this kind of issue.